### ISSUE TITLE
DecimalType comparison expressions should be valid between values with decimal types with different scales/precisions

### ISSUE THREAD

**Issue by allisonport-db** [delta-io/connectors#222]
Currently our comparison expressions fail with `IllegalArgumentException` for decimal type arguments that don’t have equal precision/scale since we check for equal data types. This should be valid (and is in Spark SQL). This impacts IN and all the binary comparison operators. 
Ex: ` new EqualTo(Literal.of(BigDecimal.valueOf(2).setScale(1)), Literal.of(BigDecimal.valueOf(2).setScale(2)))` throws an illegal argument exception for mismatched data types.

Fix: add method `DataType::equivalent(DataType other)` that defaults to equals, and override it in `DecimalType`. Update BinaryOperator and IN to use the equivalent method.

**Comment by kination** [comment]
@allisonport-db there are overrided 'equals' class in `DataType`. So you want `equivalent` to be act different with this, right?

**Comment by allisonport-db** [comment]
Hi @kination sorry for the delay.

So basically we want to add `DataType::equivalent(DataType other)` which defaults to `equals` and `DecimalType::equivalent(DataType other)` that overrides it. Then instead of using `equals` in the comparisons in our code, we will use `DataType::equivalent(DataType other)` .

Does that clarify things? If you're interested in working on this let me know and I'll go ahead and assign this issue to you.

**Comment by kination** [comment]
@allisonport-db thanks! Please let me take a look on this issue 🙏

**Comment by kination** [comment]
@allisonport-db thanks. 1 more question...
Could you let me know which build command should I use for testing the result after fix?

**Comment by kination** [comment]
@allisonport-db hello.
For test, when I run
```
Literal.of(BigDecimal.valueOf(2).setScale(1))
```

it shows
```
overloaded method value of with alternatives:
[error]   (x$1: Byte)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: java.sql.Timestamp)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: String)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Short)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Long)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Float)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Double)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: java.math.BigDecimal)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: java.sql.Date)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Array[Byte])io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Boolean)io.delta.standalone.expressions.Literal <and>
[error]   (x$1: Int)io.delta.standalone.expressions.Literal
[error]  cannot be applied to (scala.math.BigDecimal)
[error]     Literal.of(BigDecimal.valueOf(2).setScale(1))
[error]             ^
```
Is this expected error?

**Comment by allisonport-db** [comment]
I think this is because you're creating a scala BigDecimal. Our `Literal` expression supports `java.math.BigDecimal` (see here https://delta-io.github.io/connectors/latest/delta-standalone/api/java/index.html)

You can see other usages of `BigDecimal` in the ExpressionSuite (which use `BigDecimalJ`)

**Comment by kination** [comment]
@allisonport-db thanks, you were right.
As you said there are exception on 
```
new EqualTo(Literal.of(BigDecimal.valueOf(2).setScale(1)), Literal.of(BigDecimal.valueOf(2).setScale(2)))
```

So are you meaning to change comparing logic something like following?
```
Literal.of(BigDecimal.valueOf(2).setScale(1)).equivalent(Literal.of(BigDecimal.valueOf(2).setScale(2)))
```

If you could suggest some rough code format of data comparison using `equivalent`, it would be very helpful 🙏 

Thanks.

**Comment by allisonport-db** [comment]
No, we want `new EqualTo(Literal.of(BigDecimal.valueOf(2).setScale(1)), Literal.of(BigDecimal.valueOf(2).setScale(2))).eval()` to not throw an error, as it should be a valid comparison.

`equivalent` will only be used internally by our code to compare data types, which will replace any usages of `equals`. Please see https://github.com/delta-io/connectors/issues/222#issuecomment-1182481164 I think this explains the changes needed pretty well. An example of where we use `equals` to compare data types is [here](https://github.com/delta-io/connectors/blob/master/standalone/src/main/java/io/delta/standalone/expressions/BinaryOperator.java#L16).

**Comment by kination** [comment]
@allisonport-db thanks. More clear.

For now, it seems `DataType::equivalent` should compare only type name, and `DecimalType::equivalent` should compare `precision` and `scale`.

So, for `DataType::equivalent`
```
public boolean equivalent(DataType dt) {
    return getTypeName().equals(dt.getTypeName());
}
```

For `DecimalType::equivalent`
```
@Override
public boolean equivalent(DataType dt) {
    DecimalType that = (DecimalType) dt;
    return precision == that.precision && scale == that.scale;
}
```

am I understanding correct?

**Comment by kination** [comment]
@allisonport-db one more, is `DecimalType` should through error when input `DataType` is not `DecimalType`?

**Comment by allisonport-db** [comment]
(1) As mentioned in the issue description as well as my comments above, `equivalent` should just default to using `equals`. We do not need to compare type name in the `equivalent` function.

(2) Our whole goal of this fix, as mentioned in the issue description and throughout the comments, is to make comparison expressions between decimal types with different scales/precisions valid. The implementation you've provided here is just identical to our current `equals` implementation ([see here](https://github.com/delta-io/connectors/blob/master/standalone/src/main/java/io/delta/standalone/types/DecimalType.java#L78)) and does not achieve that goal. Would that implementation work for the example I gave above?
> Ex: new EqualTo(Literal.of(BigDecimal.valueOf(2).setScale(1)), Literal.of(BigDecimal.valueOf(2).setScale(2))) throws an illegal argument exception for mismatched data types.

`DecimalType.equivalent` should only check that the other DataType is also a decimal type so that the above example does not throw an error.

**Comment by allisonport-db** [comment]
If you haven't seen it, here is an example of the code that is throwing the mismatched data types error: https://github.com/delta-io/connectors/blob/master/standalone/src/main/java/io/delta/standalone/expressions/BinaryOperator.java#L16 (also linked in a comment above)

**Comment by kination** [comment]
@allisonport-db yes, I've checked that comparison logic.
https://github.com/delta-io/connectors/pull/426

Could you check the following, and whether I'm approaching correctly?

**Comment by kination** [comment]
@allisonport-db close it if there's no issue 🙏
